Proposed changes for the WebMessageObjects API to be able to inject web objects into the WebView2 content from app. (#4356)
MicrosoftEdge/WebView2Feedback: specs/WebMessageObjects.md

# Background
This is a proposal for a new API that will provide a framework for native representation of DOM
objects from page content in WebView2 and its implementation for DOM File objects. The [specific ask
from WebView2](https://github.com/MicrosoftEdge/WebView2Feedback/issues/501) is to be able get to
the paths for DOM file objects, which is not accessible to page content. We also have asks to expose
other DOM objects, including iframes, <object> objects, etc. which will be added under this
WebMessageObjects framework in future.

This API will also allow to inject DOM objects into WebView2 content constructed via the app and via
the CoreWebView2.PostWebMessage API in the other direction. This API surface needs to be compatible

> with that. (See Appendix)

Reviewer: Looks like the appendix was removed

Author: This whole sentence should be removed.

}
```

## Post a FileSystemHandle to the web content
Use this API to create a FileSystemHandle from the app and send it to the web content in WebView2.

### Web content code
```javascript
chrome.webview.addEventListener("message", function (e) {
    if (e.data.MyMessageType === "FileHandle") {
        var fileHandle = e.additionalObjects[0];
        if (fileHandle.kind === "file") {

>             // run file code

Reviewer: Would the developer be expected to check permission somewhere here?

        } else if (fileHandle.kind === "directory") {
            // run directory code
        }
    }
});
```

### App code
```cpp
wil::com_ptr<ICoreWebView2_21> webview21 =
    m_webView2.try_query<ICoreWebView2_21>();
wil::com_ptr<ICoreWebView2Environment> environment =
    appWindow->GetWebViewEnvironment();
wil::com_ptr<ICoreWebView2Environment14> environment14 = 
    environment.try_query<ICoreWebView2Environment14>();
wil::com_ptr<ICoreWebView2FileSystemHandle> fileHandle;

> CHECK_FAILURE(environment14->CreateWebFileSystemHandle(

Reviewer 2: Should be more explicit what the behavior of this is. It sounds from later comments like it's an open-if-exists operation (not fail-if-exists, not overwrite).

And what exclusive sharing?

    L"C:\\Users\\<user>\\Documents\\file.txt",
    COREWEBVIEW2_FILE_SYSTEM_HANDLE_PERMISSION_READWRITE,

>     COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND_FILE, &fileHandle));

Reviewer 3: I think it's more natural to have the Kind come first, then the Permission.

CreateWebFileSystemHandle(path, File, ReadOnly)

as opposed to

CreateWebFileSystemHandle(path, ReadOnly, File)

In fact, maybe the Kind should come ahead of the path!

CreateWebFileSystemHandle(File, path, ReadOnly)

wil::com_ptr<ICoreWebView2ObjectCollection> webObjectCollection;
IUnknown* webObjects[] = {fileHandle.get()};
CHECK_FAILURE(environment14->CreateObjectCollection(
    ARRAYSIZE(webObjects),
    webObjects,
    &webObjectCollection));
// Check the source to ensure that we are sending the message to the correct target.
wil::unique_cotaskmem_string source;
CHECK_FAILURE(webview21->get_Source(&source));

> if (std::wstring(source.get()).rfind(L"https://www.example.com/", 0) == 0) {

Reviewer 3: This seems like an awfully weird way of testing the source domain. It would reject `https://www.example.com/https://www.example.com/` which seems like it should be okay (still under the `www.example.com` domain).

Why not just `starts_with(L"https://www.example.com/")`?

    CHECK_FAILURE(webview21->PostWebMessageAsJsonWithAdditionalObjects(
        L"{ \"MyMessageType\" : \"FileHandle\" }", webObjectCollection.get()));
}
```

> ```c#

Reviewer 4: In the rendered markdown it is a little confusing having the two examples right next to each other. It initially seemed like I was reading a single example that suddenly transitioned from C++ to C#. Just add a title to the C# example to make it clearer.

// Check the source to ensure that we are sending the message to the correct target.
if (webView.CoreWebView2.Source.StartsWith("https://www.example.com/")) {
    webView.CoreWebView2.PostWebMessageAsJsonWithAdditionalObjects("{ \"MyMessageType\" : \"FileHandle\" }", new List<object>()
    {

>         webView.CoreWebView2.Environment.CreateWebFileSystemHandle(

Reviewer 2: Checking my understanding: this isn't a "handle" in the Windows sense; this call doesn't actually get a file system handle. It's really an "info", and in the web content the file will actually be opened.

  [propget] HRESULT AdditionalObjects(
      [out, retval] ICoreWebView2ObjectCollectionView** value);
}

[v1_enum]
typedef enum COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND {
  /// FileSystemHandle is for a file (i.e. FileSystemFileHandle)
  COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND_FILE,
  /// FileSystemHandle is for a directory (i.e. FileSystemDirectoryHandle)
  COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND_DIRECTORY
} COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND;

[v1_enum]
typedef enum COREWEBVIEW2_FILE_SYSTEM_HANDLE_PERMISSION {
  /// Read-only permission for FileSystemHandle
  COREWEBVIEW2_FILE_SYSTEM_HANDLE_PERMISSION_READONLY,
  /// Read and write permissions for FileSystemHandle
  COREWEBVIEW2_FILE_SYSTEM_HANDLE_PERMISSION_READWRITE
} COREWEBVIEW2_FILE_SYSTEM_HANDLE_PERMISSION;

[uuid(0ecf4d7d-bbf6-4320-930e-82ff6cf2d8dc), object, pointer_default(unique)]
interface ICoreWebView2FileSystemHandle : IUnknown {
  /// The Kind of the FileSystemHandle. It can either be a file or a directory.
  [propget] HRESULT Kind([out, retval] COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND* kind);

  /// The path to the FileSystemHandle.

>   [propget] HRESULT Path([out, retval] LPWSTR* path);

Reviewer: nit: names of properties should be `value` (+ other places)

  /// The permissions granted to the FileSystemHandle.
  [propget] HRESULT Permission([out, retval] COREWEBVIEW2_FILE_SYSTEM_HANDLE_PERMISSION* kind);
}

// Generic container of COM objects to pass to WebView2
[uuid(d2fba648-109f-401c-b701-44fd3ddfb440), object, pointer_default(unique)]
interface ICoreWebView2ObjectCollection : ICoreWebView2ObjectCollectionView {
  /// Removes the object at the specified index.
  HRESULT RemoveValueAtIndex([in] UINT32 index);

  /// Inserts the object at the specified index.
  HRESULT InsertValueAtIndex(
      [in] UINT32 index,
      [in] IUnknown* value);
}

[uuid(afae6f48-60d6-4b95-8b81-6f60d9c70f0b), object, pointer_default(unique)]
interface ICoreWebView2Environment14 : ICoreWebView2Environment13 {
  /// Create a ICoreWebView2FileSystemHandle from a path. The `path` is the path pointed by the
  /// file and must be a syntactically correct full path but it is not checked here whether it currently
  /// points to a file or a directory. If an invalid path is passed, an E_INVALIDARG will be returned

>   /// and the object will fail to create.

Reviewer 3: I assume by "full path" we mean a fully qualified path (not a relative path). Does passing a relative path also return `E_INVALIDARG`?

  ///
  /// If the `kind` is COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND_FILE, representation of a Web
  /// [FileSystemFileHandle](https://developer.mozilla.org/docs/Web/API/FileSystemFileHandle)
  /// will be created. Once the object is passed to web content, if the content is attempting a read,
  /// the file must be existing and available to read similar to a file chosen by
  /// [open file picker](https://developer.mozilla.org/docs/Web/API/Window/showOpenFilePicker),
  /// otherwise the read operation will throw a DOM exception. For write operations the file does not
  /// need to exist as `FileSystemFileHandle` will behave more like a file path chosen by 
  /// [save file picker](https://developer.mozilla.org/docs/Web/API/Window/showSaveFilePicker)
  /// and will create or overwrite the file, but the parent directory structure pointed 
  /// by the file must exist and an existing file must be available to write and delete

>   /// or the write operation will throw a DOM exception.

Reviewer 3: Can CoreWebView2 be used from UWP? (I assume not. Otherwise we have even bigger problems with paths and security.)

  ///
  /// If the `kind` is COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND_DIRECTORY, representation of a
  /// FileSystemDirectoryHandle will be created. The path must point to a directory
  /// as if it was chosen via
  /// [directory picker](https://developer.mozilla.org/docs/Web/API/Window/showDirectoryPicker)
  /// when it is accessed by web content otherwise the IO operation will throw a DOM exception.
  /// 
  /// In either case, the app must have the same permissions to the path it
  /// wishes to give the web content to read/write the file/directory due to [process

>   /// model of WebView2](https://learn.microsoft.com/microsoft-edge/webview2/concepts/process-model).

Reviewer 3: > The app must have the same permissions to the path...

Does this "same permission" honor impersonation (transfer impersonation to the WebView process so it can access the file)? I assume not, but we should say so.

  /// Note that the only validation done in this function is that the path is a valid
  /// full path and otherwise the function will succeed.
  /// Any other state validation will be done when this handle is accessed from web content
  /// and will cause DOM exceptions if access operations fail.
  HRESULT CreateWebFileSystemHandle(
      [in] LPCWSTR path,
      [in] COREWEBVIEW2_FILE_SYSTEM_HANDLE_PERMISSION permission,
      [in] COREWEBVIEW2_FILE_SYSTEM_HANDLE_KIND kind,
      [out, retval] ICoreWebView2FileSystemHandle** fileSystemHandle);

>   /// Create a ICoreWebView2File from a file path. The object created is a

Reviewer 2: I can't find the definition of `ICoreWebView2File`?

  /// representation of a DOM [File](https://developer.mozilla.org/docs/Web/API/File)
  /// object.
  /// The `path` is the full path pointed to the file. It must be a valid file 
  /// that the app has read access for and is available to read or creation will fail

>   /// with E_INVALIDARG.

Reviewer 3: It appears that this is different from CreateWebFileSystemHandle in that it validates read access immediately rather than on demand. Why the inconsistency?

There is a footnote that calls out the inconsistency but doesn't explain why.

  /// (See Footnote 1)
  HRESULT CreateWebFile(
      [in] LPCWSTR path,

>       [out, retval] ICoreWebView2File** file);

Reviewer 3: No sample for this API

  /// Create a generic object collection. 
  HRESULT CreateObjectCollection(
      [in] UINT32 length,
      [in, size_is(length)] IUnknown** items,
      [out, retval] ICoreWebView2ObjectCollection** objectCollection);
}

[uuid(5220d097-c583-4115-8e08-1af2100338c2), object, pointer_default(unique)]
interface ICoreWebView2_24 : ICoreWebView2_23 {
  /// Same as PostWebMessageAsJson, but also has support for posting DOM objects
  /// to page content. The `additionalObjects` property in the DOM MessageEvent
  /// fired on the page content is a ArrayLike list of
  /// DOM objects that can be posted to the web content. Currently these can be
  /// the following types and `null`:
  /// | Win32             | DOM type    |
  /// |-------------------|-------------|
  /// | ICoreWebView2File | [File](https://developer.mozilla.org/docs/Web/API/File) |
  /// | ICoreWebView2FileSystemHandle | [FileSystemHandle](https://developer.mozilla.org/docs/Web/API/FileSystemHandle) |
  /// | nullptr           | null        |
  /// The objects are posted the to web content following the structured-clone
  /// semantics, meaning only objects that can be cloned can be posted.

>   /// They will also behave as in they are created on the web content they are

Reviewer 3: ```suggestion
  /// They will also behave as if they had been created by the web content they are

```

  /// posted to. For example, if a FileSystemFileHandle is posted to a web content

>   /// it can only be transferred via postMessage to other web content with the same origin.

Reviewer 3: Is there anything specific about files here? It seems that postMessage in general follows the same origin policy, unrelated to additional objects.

  HRESULT PostWebMessageAsJsonWithAdditionalObjects([in] LPCWSTR webMessageAsJson, [in] ICoreWebView2ObjectCollectionView* additionalObjects);
}
```

```c#
namespace Microsoft.Web.WebView2.Core
{
    enum CoreWebView2FileSystemHandlePermission
    {
        /// Read-only permission for FileSystemHandle

>         CoreWebView2FileSystemHandlePermissionReadonly = 0,

Reviewer 4: In WinRT, the values of an enum do not need to be prefixed with the name of the enum. i.e. this could be 
```
enum CoreWebView2FileSystemHandlePermission
{
    Readonly,
    Readwrite
}
```

Similarly for CoreWebView2FileSystemHandleKind.

Reviewer 4: And `Readonly` and `Readwrite` should be camel case:
```
ReadOnly
ReadWrite
```

Reviewer 3: Indeed, the earlier C# sample used the shorter names with the correct capitalization.

(PS: PascalCase is first-letter-of-each-word. camelCase is lowercase first letter, because camels have a hump in the middle.)

        /// Read and write permissions for FileSystemHandle
        CoreWebView2FileSystemHandlePermissionReadwrite = 1,
    };
    enum CoreWebView2FileSystemHandleKind
    {
        /// FileSystemHandle is for a file (i.e. FileSystemFileHandle)
        CoreWebView2FileSystemHandleKindFile = 0,
        /// FileSystemHandle is for a directory (i.e. FileSystemDirectoryHandle)

>         CoreWebView2FileSystemHandleKindDirectory = 1,

Reviewer 3: Same here. Samples assume that these are `File` and `Directory` without the enum name prefix.

        /// Any DOM object type that can be natively representable that has been passed in to
        /// `additionalObjects` parameter will be accessible here.
        /// Currently a WebMessage object can be the following type:
        /// $net$
        /// - System.IO.FileInfo
        /// $end$
        /// $winrt$

>         /// - Windows.Storage.StorageFile

Reviewer 3: Wait, are we removing support for these types and replacing them with the two new types?

            IVectorView<Object> AdditionalObjects { get; };
        }
    }

    runtimeclass CoreWebView2
    {
        ...
        /// Same as PostWebMessageAsJson, but also has support for posting DOM objects
        /// to page content. The `additionalObjects` property in the DOM MessageEvent
        /// fired on the page content is a ArrayLike list of
        /// DOM objects that can be posted to the web content. Currently the
        /// following types can be posted to DOM:
        /// | .NET/WinRT       | DOM type    |
        /// |------------------|-------------|
        /// | CoreWebView2File | [File](https://developer.mozilla.org/docs/Web/API/File] |
        /// | CoreWebView2FileSystemHandle | [FileSystemHandle](https://developer.mozilla.org/docs/Web/API/FileSystemHandle) |
        /// | null             | null        |
        /// The objects are posted the to web content following the structured-clone
        /// semantics, meaning only objects that can be cloned can be posted.

>         /// They will also behave as in they are created on the web content they are

Reviewer 3: ```suggestion
        /// They will also behave as if they had been created by the web content they are

```

        /// posted to. For example, if a FileSystemFileHandle is posted to a web content
        /// it can only be posted via postMessage to other web content with the same origin.

>         PostWebMessageAsJsonWithAdditionalObjects(string webMessageAsJson, IVectorView<Object> additionalObjects);

Reviewer 4: Should this have an `[overload("PostWebMessageAsJson")]` so that it is projected with a cleaner name?

Reviewer 3: Or maybe it's named this way to align with the JS API postMessageWithAdditionalObjects, which is itself not an overload because there is a pre-existing overload on the JS side that takes a transferList.

            /// and will create or overwrite the file, but the parent directory structure pointed 
            /// by the file must exist and an existing file must be available to write and delete
            /// or the write operation will throw a DOM exception.
            ///
            /// If the `kind` is CoreWebView2FileSystemHandleKind.Directory, representation of a
            /// FileSystemDirectoryHandle will be created. The path must point to a directory
            /// as if it was chosen via
            /// [directory picker](https://developer.mozilla.org/docs/Web/API/Window/showDirectoryPicker)
            /// when it is accessed by web content otherwise the IO operation will throw a DOM exception.
            /// 
            /// In either case, the app must have the same permissions to the path it
            /// wishes to give the web content to read/write the file/directory due to [process
            /// model of WebView2](https://learn.microsoft.com/microsoft-edge/webview2/concepts/process-model).
            /// Note that the only validation done in this function is that the path is a valid

>             /// full path and otherwise the function will succeed.

Reviewer 2: It can't also validate the file vs directory kind? (Actually, could it calculate that automatically?)

Reviewer 5: No change. It doesn't perform disk IO

            /// Any other state validation will be done when this handle is accessed from web content
            /// and will cause DOM exceptions if access operations fail.

>             CoreWebView2FileSystemHandle CreateWebFileSystemHandle(String Path, CoreWebView2FileSystemHandlePermission Permission, CoreWebView2FileSystemHandleKind Kind);

Reviewer 2: ```suggestion
            CoreWebView2FileSystemHandle CreateWebFileSystemHandle(String path, CoreWebView2FileSystemHandlePermission permission, CoreWebView2FileSystemHandleKind kind);

```

Reviewer 5: Please fix casing.

            /// Create a `CoreWebView2File` from a file path. The object created is a
            /// representation of a DOM [File](https://developer.mozilla.org/docs/Web/API/File)
            /// object.
            /// The `path` is the full path pointed to the file. It must be a valid file 
            /// that the app has read access for and is available to read or this will throw
            /// InvalidArgumentException.
            /// (See Footnote 1)

>             CoreWebView2File CreateWebFile(String Path);

Reviewer 2: Confused about what the difference is in this API between a file and a handle.

Reviewer 5: See above.

     */
    postMessageWithAdditionalObjects(message: any, additionalObjects: ArrayLike<any>) : void;
}

interface WebViewMessageEvent extends MessageEvent {
    /**
     * The source of the event is the `chrome.webview` object.
     */
    source: WebView;

    /**
     * Additional DOM objects that are sent via PostJSONMessageWithAdditionalObjects.
     * These objects can be of following types:
     * | DOM       | Win32       | .NET/WinRT |
     * |-----------|-------------|------------|
     * | [File](https://developer.mozilla.org/docs/Web/API/File] | ICoreWebView2File | CoreWebView2File |
     * | [FileSystemHandle](https://developer.mozilla.org/docs/Web/API/FileSystemHandle) | ICoreWebView2FileSystemHandle | CoreWebView2FileSystemHandle |
     * | null      | nullptr     | null       |
     * | undefined | nullptr     | null       |
     */
    additionalObjects: ArrayLike<any>
}

// Matching `interface MessagePort` from lib.dom.d.ts for messaging APIs
/**
 * Events of the `WebView` interface.
 */
interface WebViewEventMap {
    "message": WebViewMessageEvent;

>     "sharedbufferreceived": SharedBufferReceivedEvent;

Reviewer 3: Not sure why this is here in this spec.

Reviewer 5: This is unrelated to this feature. Please go back to '...'
